### ISSUE TITLE
RealmBaseAdapter change listeners scope is too wide

### ISSUE THREAD

**Issue by JamesMcIntosh** [realm/realm-android-adapters#3]
I've been having a problem with the BaseAdapter firing when I pop a widget in the back stack because of the AutoCompleteTextView.PopupDataSetObserver runnable.

While looking into what was causing the update I noticed that the listeners being registered are done so as global realm listeners. i.e.

https://github.com/realm/realm-android-adapters/blob/master/adapters/src/main/java/io/realm/RealmBaseAdapter.java

```
realmResults.realm.handlerController.addChangeListenerAsWeakReference(listener);
```

Would it be better to just register the listeners directly onto the realmResults itself then listening to possibly erroneous updates of other elements in the realm which do not affect the query but will cause it to update?

```
realmResults.addChangeListener(listener);
```

I can see that there is potentially a need for it to be registered globally if the information displayed from the adapter is transitively looks up data from elsewhere but I would expect that you would register a listener for this case explicitly.

**Comment by cmelchior** [comment]
Yes, it is correct that the current implementation is a bit too broad. The reason is that BaseListAdapter doesn't have a lifecycle callback for when the Adapter is removed from a ListView, which means we have no hook for unregistering our listener. So we resort to using weak references. Unfortunately that part of the code hasn't been improved together with the rest of our collection listeners. It is on our radar though.

If you want to work around it. You can adopt our `RealmBaseAdapter` class and add a `unregister` method that explicitly removes the change listener. You will need to call that method explicitly from your Activity when your ListView goes is removed (probably on onDestroy), to avoid leaks.

**Comment by nhachicha** [comment]
@cmelchior isn't this already solved by https://github.com/realm/realm-android-adapters/blob/425e2bb8f025189bce57ab53be4f5866151a1c78/adapters/src/main/java/io/realm/RealmBaseAdapter.java#L59 ?

**Comment by Zhuinden** [comment]
Isn't this fixed by Realm 3.0.0 and https://github.com/realm/realm-java/pull/3834 with https://github.com/realm/realm-java/pull/4216 ?

**Comment by cmelchior** [comment]
Yes it is. Closing
